File: services/ai-detection/engines/risk_radar.py
"""
Risk Radar Engine
8-vector supply chain risk assessment + heatmap.
Ported from server/engines/risk-radar.js
"""

from __future__ import annotations

from datetime import datetime, timezone, timedelta
from typing import Any
import math


HIGH_RISK_REGIONS = {"CN": 35, "RU": 45, "IN": 20, "KR": 10, "TH": 15}
VECTOR_WEIGHTS = {
    "partner_risk": 0.20, "geographic_risk": 0.10, "route_risk": 0.15,
    "financial_risk": 0.15, "compliance_risk": 0.10, "cyber_risk": 0.10,
    "environmental_risk": 0.05, "supply_disruption": 0.15,
}


def _level(score: float) -> str:
    return "high" if score > 60 else ("medium" if score > 30 else "low")


# ─── Vector Assessors ──────────────────────────────────────────

def assess_partner_risk(partners: list, violations: list) -> dict:
    if not partners:
        return {"score": 0, "level": "low", "details": {}}
    n = len(partners)
    kyc_failed = sum(1 for p in partners if p.get("kyc_status") in ("failed", "pending"))
    low_trust = sum(1 for p in partners if p.get("trust_score", 50) < 50)
    viol = len(violations)
    score = min(100, kyc_failed / n * 40 + low_trust / n * 30 + min(viol, 10) * 3)
    return {
        "score": round(score, 1), "level": _level(score),
        "details": {
            "total_partners": n, "kyc_incomplete": kyc_failed,
            "low_trust_partners": low_trust, "sla_violations": viol,
            "avg_trust_score": round(sum(p.get("trust_score", 50) for p in partners) / n),
        },
    }


def assess_geographic_risk(partners: list, shipments: list) -> dict:
    conc: dict[str, int] = {}
    for p in partners:
        c = p.get("country", "XX")
        conc[c] = conc.get(c, 0) + 1
    total = len(partners) or 1
    hhi = sum((cnt / total) ** 2 for cnt in conc.values()) * 10000
    country_risk = sum(HIGH_RISK_REGIONS.get(p.get("country", ""), 5) for p in partners) / total if total else 0
    score = min(100, hhi / 100 + country_risk)
    return {
        "score": round(score, 1), "level": _level(score),
        "details": {
            "herfindahl_index": round(hhi),
            "concentration_warning": "High concentration — diversification needed" if hhi > 5000 else "Adequate diversification",
            "country_distribution": conc,
            "avg_country_risk": round(country_risk, 1),
        },
    }


def assess_route_risk(shipments: list) -> dict:
    if not shipments:
        return {"score": 0, "level": "low", "details": {}}
    late = [s for s in shipments if s.get("actual_delivery") and s.get("estimated_delivery") and s["actual_delivery"] > s["estimated_delivery"]]
    late_ids = {s.get("id") for s in late}
    in_transit = [s for s in shipments if s.get("status") in ("in_transit", "pending")]
    lr = len(late) / len(shipments)
    avg_delay = 0.0
    if late:
        delays = []
        for s in late:
            try:
                d1 = datetime.fromisoformat(str(s["actual_delivery"]).replace("Z", "+00:00"))
                d2 = datetime.fromisoformat(str(s["estimated_delivery"]).replace("Z", "+00:00"))
                delays.append((d1 - d2).total_seconds() / 3600)
            except Exception:
                pass
        avg_delay = sum(delays) / len(delays) if delays else 0.0

    carrier_stats: dict[str, dict] = {}
    for s in shipments:
        c = s.get("carrier", "Unknown")
        carrier_stats.setdefault(c, {"total": 0, "late": 0})
        carrier_stats[c]["total"] += 1
        if s.get("id") in late_ids:
            carrier_stats[c]["late"] += 1

    score = min(100, lr * 80 + min(avg_delay / 48, 1) * 20)
    return {
        "score": round(score, 1), "level": _level(score),
        "details": {
            "total_shipments": len(shipments), "late_shipments": len(late),
            "on_time_rate": f"{round((1 - lr) * 100)}%",
            "avg_delay_hours": round(avg_delay, 1), "in_transit": len(in_transit),
            "carrier_performance": [
                {"carrier": c, "total": s["total"], "late": s["late"], "reliability": f"{round((1 - s['late'] / s['total']) * 100)}%"}
                for c, s in carrier_stats.items()
            ],
        },
    }


def assess_financial_risk(leaks: list, violations: list) -> dict:
    n = len(leaks)
    price_dev = sum(abs(l["listing_price"] - l["authorized_price"]) / l["authorized_price"] for l in leaks if l.get("authorized_price") and l.get("listing_price"))
    penalty = sum(v.get("penalty_amount", 0) for v in violations)
    score = min(100, min(n * 5, 40) + (price_dev / n * 30 if n else 0) + min(penalty / 1000, 30))
    return {
        "score": round(score, 1), "level": _level(score),
        "details": {
            "leak_alerts": n,
            "avg_price_deviation": f"{round(price_dev / n * 100)}%" if n else "0%",
            "total_penalties": penalty,
            "gray_market_risk": "elevated" if n > 3 else "normal",
        },
    }


def assess_compliance_risk(certifications: list) -> dict:
    if not certifications:
        return {"score": 20, "level": "low", "details": {"message": "No certifications tracked"}}
    now = datetime.now(timezone.utc)
    expired = 0
    expiring = 0
    cutoff = now + timedelta(days=30)
    for c in certifications:
        raw = c.get("expiry_date")
        if not raw:
            continue
        try:
            dt = datetime.fromisoformat(str(raw).replace("Z", "+00:00"))
        except Exception:
            continue
        if dt < now:
            expired += 1
        elif dt < cutoff:
            expiring += 1
    n = len(certifications)
    score = min(100, expired / n * 60 + expiring / n * 20 + 5)
    return {
        "score": round(score, 1), "level": _level(score),
        "details": {"total_certifications": n, "expired": expired, "expiring_30_days": expiring, "active": n - expired},
    }


def assess_cyber_risk(partners: list, alerts: list) -> dict:
    no_api = sum(1 for p in partners if not p.get("api_key"))
    fraud_a = sum(1 for a in alerts if a.get("alert_type") == "STATISTICAL_ANOMALY" or a.get("severity") == "high")
    score = min(100, no_api / max(len(partners), 1) * 30 + min(fraud_a * 8, 40) + 10)
    return {
        "score": round(score, 1), "level": _level(score),
        "details": {"partners_without_api_auth": no_api, "anomaly_alerts": fraud_a, "data_integrity": "blockchain_sealed"},
    }


def assess_environmental_risk(sustainability: list) -> dict:
    if not sustainability:
        return {"score": 30, "level": "medium", "details": {"message": "No ESG assessments — risk unmeasured"}}
    avg = sum(s.get("overall_score", 50) for s in sustainability) / len(sustainability)
    low = sum(1 for s in sustainability if s.get("overall_score", 50) < 50)
    score = round(max(0, 100 - avg))
    return {
        "score": score, "level": _level(score),
        "details": {"avg_esg_score": round(avg, 1), "products_assessed": len(sustainability), "low_performers": low, "carbon_risk": "elevated" if avg < 60 else "acceptable"},
    }


def assess_supply_disruption(inventory: list, partners: list, shipments: list) -> dict:
    suppliers = sum(1 for p in partners if p.get("type") in ("oem", "manufacturer"))
    ssr = 80 if suppliers <= 1 else (40 if suppliers == 2 else 10)
    low = sum(1 for i in inventory if i.get("quantity", 0) <= i.get("min_stock", 10))
    sr = low / len(inventory) if inventory else 0
    now_ts = datetime.now(timezone.utc)
    stuck = 0
    for s in shipments:
        if s.get("status") == "in_transit" and s.get("created_at"):
            try:
                c = datetime.fromisoformat(str(s["created_at"]).replace("Z", "+00:00"))
                if (now_ts - c).days > 14:
                    stuck += 1
            except Exception:
                pass
    score = min(100, ssr * 0.35 + sr * 100 * 0.35 + min(stuck * 10, 30) * 0.30)
    return {
        "score": round(score, 1), "level": _level(score),
        "details": {"supplier_diversity": suppliers, "single_source_dependencies": suppliers <= 1, "low_stock_items": low, "total_inventory_items": len(inventory), "stuck_shipments": stuck},
    }


# ─── Main Functions ───────────────────────────────────────────

def compute_radar(data: dict[str, Any] | None = None) -> dict:
    """Compute full risk radar — all 8 vectors."""
    data = data or {}
    partners = data.get("partners", [])
    shipments = data.get("shipments", [])
    violations = data.get("violations", [])
    leaks = data.get("leaks", [])
    alerts = data.get("alerts", [])
    inventory = data.get("inventory", [])
    certs = data.get("certifications", [])
    sust = data.get("sustainability", [])

    vectors = {
        "partner_risk": assess_partner_risk(partners, violations),
        "geographic_risk": assess_geographic_risk(partners, shipments),
        "route_risk": assess_route_risk(shipments),
        "financial_risk": assess_financial_risk(leaks, violations),
        "compliance_risk": assess_compliance_risk(certs),
        "cyber_risk": assess_cyber_risk(partners, alerts),
        "environmental_risk": assess_environmental_risk(sust),
        "supply_disruption": assess_supply_disruption(inventory, partners, shipments),
    }

    overall = sum(vectors[k]["score"] * w for k, w in VECTOR_WEIGHTS.items())

    return {
        "overall_threat_index": round(overall, 1),
        "threat_level": "critical" if overall > 70 else ("high" if overall > 50 else ("medium" if overall > 30 else "low")),
        "vectors": vectors,
        "updated_at": datetime.now(timezone.utc).isoformat(),
    }


def generate_heatmap(partners: list, shipments: list, leaks: list) -> list[dict]:
    """Generate risk heatmap by region."""
    region_map: dict[str, dict] = {}
    for p in partners:
        r = p.get("country") or p.get("region") or "Unknown"
        region_map.setdefault(r, {"partners": 0, "risk_score": 0, "leaks": 0, "shipments": 0})
        region_map[r]["partners"] += 1
        region_map[r]["risk_score"] += 100 - p.get("trust_score", 50)

    for l in leaks:
        r = l.get("region_detected", "Unknown")
        region_map.setdefault(r, {"partners": 0, "risk_score": 0, "leaks": 0, "shipments": 0})
        region_map[r]["leaks"] += 1
        region_map[r]["risk_score"] += (l.get("risk_score", 0.5)) * 20

    result = []
    for region, d in region_map.items():
        heat = round(min(100, d["risk_score"] / max(d["partners"], 1)))
        result.append({
            "region": region, "heat_score": heat, "partners": d["partners"],
            "leak_alerts": d["leaks"],
            "risk_level": "hot" if heat > 50 else ("warm" if heat > 25 else "cool"),
        })
    return sorted(result, key=lambda x: x["heat_score"], reverse=True)

File: server/engines/intelligence/intelligence/risk-radar.js
/**
 * TrustChecker Risk Radar Engine
 * Multi-dimensional supply chain risk assessment across 8 vectors
 * Aggregates partner, geographic, route, financial, compliance, cyber,
 * environmental, and supply disruption risks into unified threat index
 */

class RiskRadar {
    /**
     * Compute full risk radar — all 8 dimensions
     */
    computeRadar(data = {}) {
        const { partners = [], shipments = [], violations = [], leaks = [], alerts = [], inventory = [], certifications = [], sustainability = [] } = data;

        const vectors = {
            partner_risk: this.assessPartnerRisk(partners, violations),
            geographic_risk: this.assessGeographicRisk(partners, shipments),
            route_risk: this.assessRouteRisk(shipments),
            financial_risk: this.assessFinancialRisk(leaks, violations),
            compliance_risk: this.assessComplianceRisk(certifications),
            cyber_risk: this.assessCyberRisk(partners, alerts),
            environmental_risk: this.assessEnvironmentalRisk(sustainability),
            supply_disruption: this.assessSupplyDisruption(inventory, partners, shipments)
        };

        // Overall threat index (0-100, higher = more risky)
        const weights = { partner_risk: 0.2, geographic_risk: 0.1, route_risk: 0.15, financial_risk: 0.15, compliance_risk: 0.1, cyber_risk: 0.1, environmental_risk: 0.05, supply_disruption: 0.15 };
        let overallScore = 0;
        for (const [key, weight] of Object.entries(weights)) {
            overallScore += vectors[key].score * weight;
        }

        return {
            overall_threat_index: Math.round(overallScore * 10) / 10,
            threat_level: overallScore > 70 ? 'critical' : overallScore > 50 ? 'high' : overallScore > 30 ? 'medium' : 'low',
            vectors,
            updated_at: new Date().toISOString()
        };
    }

    // ─── Vector 1: Partner Risk ──────────────────────────────────────────────
    assessPartnerRisk(partners, violations) {
        if (partners.length === 0) return { score: 0, level: 'low', details: {} };

        const kycFailed = partners.filter(p => p.kyc_status === 'failed' || p.kyc_status === 'pending').length;
        const lowTrust = partners.filter(p => (p.trust_score || 50) < 50).length;
        const violationCount = violations.length;

        const score = Math.min(100,
            (kycFailed / partners.length * 40) +
            (lowTrust / partners.length * 30) +
            (Math.min(violationCount, 10) * 3)
        );

        return {
            score: Math.round(score * 10) / 10,
            level: score > 60 ? 'high' : score > 30 ? 'medium' : 'low',
            details: {
                total_partners: partners.length,
                kyc_incomplete: kycFailed,
                low_trust_partners: lowTrust,
                sla_violations: violationCount,
                avg_trust_score: Math.round(partners.reduce((s, p) => s + (p.trust_score || 50), 0) / partners.length)
            }
        };
    }

    // ─── Vector 2: Geographic Risk ───────────────────────────────────────────
    assessGeographicRisk(partners, shipments) {
        const highRiskRegions = { 'CN': 35, 'RU': 45, 'IN': 20, 'KR': 10, 'TH': 15 };
        const concentrationMap = {};

        partners.forEach(p => {
            const country = p.country || 'XX';
            concentrationMap[country] = (concentrationMap[country] || 0) + 1;
        });

        // Concentration risk: Herfindahl-Hirschman Index
        const totalPartners = partners.length || 1;
        let hhi = 0;
        for (const count of Object.values(concentrationMap)) {
            const share = count / totalPartners;
            hhi += share * share;
        }
        hhi = hhi * 10000; // Normalize to 0-10000

        // Country risk
        let countryRisk = 0;
        partners.forEach(p => {
            countryRisk += highRiskRegions[p.country] || 5;
        });
        countryRisk = totalPartners > 0 ? countryRisk / totalPartners : 0;

        const score = Math.min(100, hhi / 100 + countryRisk);

        return {
            score: Math.round(score * 10) / 10,
            level: score > 60 ? 'high' : score > 30 ? 'medium' : 'low',
            details: {
                herfindahl_index: Math.round(hhi),
                concentration_warning: hhi > 5000 ? 'High concentration — diversification needed' : 'Adequate diversification',
                country_distribution: concentrationMap,
                avg_country_risk: Math.round(countryRisk * 10) / 10
            }
        };
    }

    // ─── Vector 3: Route Risk ────────────────────────────────────────────────
    assessRouteRisk(shipments) {
        if (shipments.length === 0) return { score: 0, level: 'low', details: {} };

        const lateShipments = shipments.filter(s => {
            if (!s.actual_delivery || !s.estimated_delivery) return false;
            return new Date(s.actual_delivery) > new Date(s.estimated_delivery);
        });

        const inTransit = shipments.filter(s => s.status === 'in_transit' || s.status === 'pending');
        const lateRate = lateShipments.length / shipments.length;
        const avgDelay = lateShipments.length > 0
            ? lateShipments.reduce((s, sh) => s + (new Date(sh.actual_delivery) - new Date(sh.estimated_delivery)) / 3600000, 0) / lateShipments.length
            : 0;

        const carrierStats = {};
        shipments.forEach(s => {
            const c = s.carrier || 'Unknown';
            if (!carrierStats[c]) carrierStats[c] = { total: 0, late: 0 };
            carrierStats[c].total++;
            if (lateShipments.find(ls => ls.id === s.id)) carrierStats[c].late++;
        });

        const score = Math.min(100, lateRate * 80 + Math.min(avgDelay / 48, 1) * 20);

        return {
            score: Math.round(score * 10) / 10,
            level: score > 60 ? 'high' : score > 30 ? 'medium' : 'low',
            details: {
                total_shipments: shipments.length,
                late_shipments: lateShipments.length,
                on_time_rate: Math.round((1 - lateRate) * 100) + '%',
                avg_delay_hours: Math.round(avgDelay * 10) / 10,
                in_transit: inTransit.length,
                carrier_performance: Object.entries(carrierStats).map(([c, s]) => ({
                    carrier: c, total: s.total, late: s.late,
                    reliability: Math.round((1 - s.late / s.total) * 100) + '%'
                }))
            }
        };
    }

    // ─── Vector 4: Financial Risk ────────────────────────────────────────────
    assessFinancialRisk(leaks, violations) {
        const leakCount = leaks.length;
        const priceDeviation = leaks.reduce((s, l) => {
            if (l.authorized_price && l.listing_price) {
                return s + Math.abs(l.listing_price - l.authorized_price) / l.authorized_price;
            }
            return s;
        }, 0);

        const penaltyTotal = violations.reduce((s, v) => s + (v.penalty_amount || 0), 0);
        const score = Math.min(100,
            Math.min(leakCount * 5, 40) +
            (leakCount > 0 ? (priceDeviation / leakCount) * 30 : 0) +
            Math.min(penaltyTotal / 1000, 30)
        );

        return {
            score: Math.round(score * 10) / 10,
            level: score > 60 ? 'high' : score > 30 ? 'medium' : 'low',
            details: {
                leak_alerts: leakCount,
                avg_price_deviation: leakCount > 0 ? Math.round(priceDeviation / leakCount * 100) + '%' : '0%',
                total_penalties: penaltyTotal,
                gray_market_risk: leakCount > 3 ? 'elevated' : 'normal'
            }
        };
    }

    // ─── Vector 5: Compliance Risk ───────────────────────────────────────────
    assessComplianceRisk(certifications) {
        if (certifications.length === 0) return { score: 20, level: 'low', details: { message: 'No certifications tracked' } };

        const expired = certifications.filter(c => c.expiry_date && new Date(c.expiry_date) < new Date()).length;
        const expiringSoon = certifications.filter(c => {
            if (!c.expiry_date) return false;
            const diff = new Date(c.expiry_date) - new Date();
            return diff > 0 && diff < 30 * 24 * 3600 * 1000;
        }).length;

        const score = Math.min(100, (expired / certifications.length * 60) + (expiringSoon / certifications.length * 20) + 5);

        return {
            score: Math.round(score * 10) / 10,
            level: score > 60 ? 'high' : score > 30 ? 'medium' : 'low',
            details: {
                total_certifications: certifications.length,
                expired: expired,
                expiring_30_days: expiringSoon,
                active: certifications.length - expired
            }
        };
    }

    // ─── Vector 6: Cyber Risk ────────────────────────────────────────────────
    assessCyberRisk(partners, alerts) {
        const noApiKey = partners.filter(p => !p.api_key).length;
        const fraudAlerts = alerts.filter(a => a.alert_type === 'STATISTICAL_ANOMALY' || a.severity === 'high').length;

        const score = Math.min(100,
            (noApiKey / Math.max(partners.length, 1) * 30) +
            Math.min(fraudAlerts * 8, 40) + 10
        );

        return {
            score: Math.round(score * 10) / 10,
            level: score > 60 ? 'high' : score > 30 ? 'medium' : 'low',
            details: {
                partners_without_api_auth: noApiKey,
                anomaly_alerts: fraudAlerts,
                data_integrity: 'blockchain_sealed'
            }
        };
    }

    // ─── Vector 7: Environmental Risk ────────────────────────────────────────
    assessEnvironmentalRisk(sustainability) {
        if (sustainability.length === 0) return { score: 30, level: 'medium', details: { message: 'No ESG assessments — risk unmeasured' } };

        const avgScore = sustainability.reduce((s, ss) => s + (ss.overall_score || 50), 0) / sustainability.length;
        const lowPerformers = sustainability.filter(s => (s.overall_score || 50) < 50).length;

        const score = Math.round(Math.max(0, 100 - avgScore));

        return {
            score,
            level: score > 60 ? 'high' : score > 30 ? 'medium' : 'low',
            details: {
                avg_esg_score: Math.round(avgScore * 10) / 10,
                products_assessed: sustainability.length,
                low_performers: lowPerformers,
                carbon_risk: avgScore < 60 ? 'elevated' : 'acceptable'
            }
        };
    }

    // ─── Vector 8: Supply Disruption ─────────────────────────────────────────
    assessSupplyDisruption(inventory, partners, shipments) {
        // Single-source dependency check
        const supplierCount = partners.filter(p => p.type === 'oem' || p.type === 'manufacturer').length;
        const singleSourceRisk = supplierCount <= 1 ? 80 : supplierCount === 2 ? 40 : 10;

        // Inventory health
        const lowStock = inventory.filter(i => i.quantity <= (i.min_stock || 10)).length;
        const stockRatio = inventory.length > 0 ? lowStock / inventory.length : 0;

        // Pending shipments that might be stuck
        const stuckShipments = shipments.filter(s => {
            if (s.status !== 'in_transit') return false;
            const daysInTransit = (Date.now() - new Date(s.created_at).getTime()) / (24 * 3600 * 1000);
            return daysInTransit > 14;
        }).length;

        const score = Math.min(100,
            singleSourceRisk * 0.35 +
            stockRatio * 100 * 0.35 +
            Math.min(stuckShipments * 10, 30) * 0.30
        );

        return {
            score: Math.round(score * 10) / 10,
            level: score > 60 ? 'high' : score > 30 ? 'medium' : 'low',
            details: {
                supplier_diversity: supplierCount,
                single_source_dependencies: supplierCount <= 1,
                low_stock_items: lowStock,
                total_inventory_items: inventory.length,
                stuck_shipments: stuckShipments
            }
        };
    }

    /**
     * Generate risk heatmap by region
     */
    generateHeatmap(partners, shipments, leaks) {
        const regionMap = {};

        partners.forEach(p => {
            const region = p.country || p.region || 'Unknown';
            if (!regionMap[region]) regionMap[region] = { partners: 0, risk_score: 0, leaks: 0, shipments: 0 };
            regionMap[region].partners++;
            regionMap[region].risk_score += (100 - (p.trust_score || 50));
        });

        leaks.forEach(l => {
            const region = l.region_detected || 'Unknown';
            if (!regionMap[region]) regionMap[region] = { partners: 0, risk_score: 0, leaks: 0, shipments: 0 };
            regionMap[region].leaks++;
            regionMap[region].risk_score += (l.risk_score || 0.5) * 20;
        });

        return Object.entries(regionMap).map(([region, data]) => ({
            region,
            heat_score: Math.round(Math.min(100, data.risk_score / Math.max(data.partners, 1))),
            partners: data.partners,
            leak_alerts: data.leaks,
            risk_level: data.risk_score / Math.max(data.partners, 1) > 50 ? 'hot' : data.risk_score / Math.max(data.partners, 1) > 25 ? 'warm' : 'cool'
        })).sort((a, b) => b.heat_score - a.heat_score);
    }
}

module.exports = new RiskRadar();
